File: Explore/2020April/Week3/C/number-of-islands.c
const char WATER = '0';
const char LAND = '1';
const char NEW = 'X';
const char USED = 'O';

void floodFill(char **grid, int ROWS, int COLS, int i, int j)
{
    if((i < 0)||(j < 0)||(i >= ROWS)||(j >= COLS)||(grid[i][j] != LAND))
        return;
    
    grid[i][j] = NEW;
    floodFill(grid, ROWS, COLS, i + 1, j);
    floodFill(grid, ROWS, COLS, i - 1, j);
    floodFill(grid, ROWS, COLS, i, j + 1);
    floodFill(grid, ROWS, COLS, i, j - 1);
}

int numIslands(char** grid, int gridSize, int* gridColSize){
    if(gridSize == 0)
        return 0;
    
    const int ROWS = gridSize;
    const int COLS = gridColSize[0];
    int numIsland = 0;
    
    for(int i = 0; i < ROWS; i++)
    {
        for(int j = 0; j < COLS; j++)
        {
            if(grid[i][j] == LAND)
            {
                floodFill(grid, ROWS, COLS, i, j);
                
                for(int x = 0; x < ROWS; x++)
                    for(int y = 0; y < COLS; y++)
                        if(grid[x][y] == NEW)
                            grid[x][y] == USED;
                
                numIsland++;
            }
        }
    }
    
    return numIsland;
}

File: Explore/2020April/Week3/Python/number-of-islands.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        if not grid:
            return 0;
        
        rows, cols = len(grid), len(grid[0])
        numIsland = 0
        
        for row in range(rows):
            for col in range(cols):
                if grid[row][col] == '1':
                    self.floodFill(grid, row, col)
                    numIsland += 1
                    
        return numIsland
    
    def floodFill(self, grid, row, col):
        if row < 0 or col < 0 or row >= len(grid) or col >= len(grid[0]) or grid[row][col] != '1':
            return 
        
        grid[row][col] = '0'
        
        self.floodFill(grid, row + 1, col)
        self.floodFill(grid, row - 1, col)
        self.floodFill(grid, row, col + 1)
        self.floodFill(grid, row, col - 1)
